hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java

```diff
@@ -0,0 +1,181 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.spi.impl;
+
+import com.hazelcast.client.config.ClientNetworkConfig;
+import com.hazelcast.client.impl.connection.AddressProvider;
+import com.hazelcast.client.properties.ClientProperty;
+import com.hazelcast.cluster.Address;
+import com.hazelcast.config.SSLConfig;
+import com.hazelcast.instance.EndpointQualifier;
+import com.hazelcast.instance.ProtocolType;
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.internal.util.AddressUtil;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.properties.HazelcastProperties;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+class TranslateToPublicAddressProvider {
+    private static final int REACHABLE_ADDRESS_TIMEOUT_MILLIS = 1000;
+    private static final int NON_REACHABLE_ADDRESS_TIMEOUT_MILLIS = 3000;
+    private static final int REACHABLE_CHECK_NUMBER = 3;
+    private static final EndpointQualifier CLIENT_PUBLIC_ENDPOINT_QUALIFIER =
+            EndpointQualifier.resolve(ProtocolType.CLIENT, "public");
+    private final ClientNetworkConfig config;
+    private final HazelcastProperties properties;
+    private final ILogger logger;
+
+    private volatile boolean translateToPublicAddress;
+
+    TranslateToPublicAddressProvider(ClientNetworkConfig config, HazelcastProperties properties, ILogger logger) {
+        this.config = config;
+        this.properties = properties;
+        this.logger = logger;
+    }
+
+    void refresh(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        translateToPublicAddress = resolve(addressProvider, members);
+    }
+
+    private boolean resolve(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        if (!(addressProvider instanceof DefaultAddressProvider)) {
+            return false;
+        }
+
+        // Default value of DISCOVERY_SPI_PUBLIC_IP_ENABLED is `null` intentionally.
+        // If DISCOVERY_SPI_PUBLIC_IP_ENABLED is not set to true/false, we don't know the intention of the user,
+        // we will try to decide if we should use private/public address automatically in that case.
+        String publicIpEnabledProperty = properties.getString(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+        if (publicIpEnabledProperty == null) {
+            SSLConfig sslConfig = config.getSSLConfig();
+            if (sslConfig != null && sslConfig.isEnabled()) {
+                if (logger.isFineEnabled()) {
```

Comment:
This check looks incorrect, as further logging is done with info level.

Reply:
Ups. It should have been fine level. I don't want to log these in info level. I think it is too internal for the user to see by default.

Reply:
I would rather use info level. This may help users and support to understand what's going on under the hood.